nano/node/fair_queue.hpp

```diff
@@ -0,0 +1,290 @@
+#pragma once
+
+#include <nano/lib/utility.hpp>
+#include <nano/node/transport/channel.hpp>
+
+#include <algorithm>
+#include <chrono>
+#include <deque>
+#include <functional>
+#include <memory>
+#include <numeric>
+#include <tuple>
+#include <utility>
+
+namespace nano
+{
+template <typename Request, typename Source>
+class fair_queue final
+{
+public:
+	/**
+	 * Holds user supplied source type(s) and an optional channel. This is used to uniquely identify and categorize the source of a request.
+	 */
+	struct origin
+	{
+		Source source;
+		std::shared_ptr<nano::transport::channel> channel;
+
+		origin (Source source, std::shared_ptr<nano::transport::channel> channel = nullptr) :
+			source{ source },
+			channel{ std::move (channel) }
+		{
+		}
+
+		bool alive () const
+		{
+			if (channel)
+			{
+				return channel->alive ();
+			}
+			// Some sources (eg. local RPC) don't have an associated channel, never remove their queue
+			return true;
+		}
+
+		auto operator<=> (origin const &) const = default;
+	};
+
+private:
+	struct entry
```

Comment:
The name entry is a bit too generic for something so important. How about origin_queue?